// testcontainers/dist/network.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StartedNetwork = exports.Network = void 0;
const docker_client_factory_1 = require("./docker-client-factory");
const uuid_1 = require("./uuid");
class Network {
    constructor(createNetworkOptions = {}, uuid = new uuid_1.RandomUuid(), dockerClientFactory = new docker_client_factory_1.DockerodeClientFactory()) {
        this.uuid = uuid;
        this.dockerClientFactory = dockerClientFactory;
        this.createNetworkOptions = Object.assign({ name: this.uuid.nextUuid(), driver: "bridge", checkDuplicate: true, internal: false, attachable: false, ingress: false, enableIPv6: false }, createNetworkOptions);
    }
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            const id = yield this.dockerClientFactory.getClient().createNetwork(this.createNetworkOptions);
            return new StartedNetwork(id, this.createNetworkOptions, this.dockerClientFactory);
        });
    }
}
exports.Network = Network;
class StartedNetwork {
    constructor(id, options, dockerClientFactory) {
        this.id = id;
        this.options = options;
        this.dockerClientFactory = dockerClientFactory;
    }
    getId() {
        return this.id;
    }
    getName() {
        return this.options.name;
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.dockerClientFactory.getClient().removeNetwork(this.id);
        });
    }
}
exports.StartedNetwork = StartedNetwork;
